fusesoc/simulator/xsim.py

```diff
@@ -9,101 +9,150 @@ class Xsim(Simulator):
+            raise OptionSectionMissing('top_module')
 
 
     def configure(self, args):
         super(Xsim, self).configure(args)
+
+        self.simcwd = os.path.join(self.work_root,
+                                   self.system.sanitized_name+'.sim',
+                                   'sim_1', 'behav')
+
         self._write_config_files()
 
+    def _has_dpi(self):
+        return (len(self.dpi_srcs) > 0)
+
     def _write_config_files(self):
-        xsim_file = 'xsim.prj'
-        f1 = open(os.path.join(self.sim_root,xsim_file),'w')
-        self.incdirs = set()
-        src_files = []
+        if self.top_module is None:
+            pr_err("No top_module set for this simulation")
+            exit(1)
+
+        ip = []         # IP descriptions (xci files)
+        constr = []     # Constraints (xdc files)
+        verilog = []    # (System) Verilog files
+        vhdl = []       # VHDL files
 
         (src_files, self.incdirs) = self._get_fileset_files(['sim', 'xsim'])
-        for src_file in src_files:
-            if src_file.file_type in ["verilogSource",
-		                      "verilogSource-95",
-		                      "verilogSource-2001"]:
-                f1.write('verilog work ' + src_file.name + '\n')
-            elif src_file.file_type in ["vhdlSource",
-                                        "vhdlSource-87",
-                                        "vhdlSource-93"]:
-                f1.write('vhdl work ' + src_file.logical_name + " " + src_file.name + '\n')
-            elif src_file.file_type in ['vhdlSource-2008']:
-                f1.write('vhdl2008 ' + src_file.logical_name + " " + src_file.name + '\n')
-            elif src_file.file_type in ["systemVerilogSource",
-                                        "systemVerilogSource-3.0",
-                                        "systemVerilogSource-3.1",
-                                        "systemVerilogSource-3.1a",
-                                        "verilogSource-2005"]:
-                f1.write('sv work ' + src_file.name + '\n')
-            else:
-                _s = "{} has unknown file type '{}'"
-                pr_warn(_s.format(src_file.name,
-                                  src_file.file_type))
-        f1.close()
-
-        tcl_file = 'xsim.tcl'
-        f2 = open(os.path.join(self.sim_root,tcl_file),'w')
-        f2.write('add_wave -radix hex /\n')
-        f2.write('run all\n')
-        f2.close()
+
+        for s in src_files:
+            if s.file_type == 'xci':
+                ip.append(s.name)
+            elif s.file_type == 'xdc':
+                constr.append(s.name)
+            elif s.file_type.startswith('verilogSource'):
+                verilog.append(s.name)
+            elif s.file_type.startswith('systemVerilogSource'):
+                verilog.append(s.name)
+            elif s.file_type.startswith('vhdlSource'):
+                vhdl.append(s.name)
+            elif s.file_type in ("CPP", "C"):
```

Comment:
I see three problems with this:
1. This means that you can add DPI source by either setting them in the DPI section or adding them with file type set to C/CPP files. Don't like to have multiple ways to do this. Given the choice, I would prefer to list them only in the filesets, since we have more control over scope and usage there.
2. I'm using the IP-XACT file types when possible. The correct types would be cSource and cppSource
3. I can't see a way to create multiple DPI libraries here. I know at least one design I'm working on packaging that uses multiple DPI libraries. Might be ok to only have one for a first implementation though. Need to think about that